sys/kerncore/src/lib.rs

```diff
@@ -0,0 +1,389 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Allow std-y things to be used in test. Note that this attribute is a bit of a
+// trap for the programmer, because rust-analyzer by default seems to build
+// things with test set. This means it's easy to introduce code incompatible
+// with no_std without your editor hassling you about it. Beware.
+#![cfg_attr(not(test), no_std)]
+
+/// Describes types that act as "slices" (in the very abstract sense) referenced
+/// by tasks in syscalls.
+///
+/// This is not the same thing as a Rust slice in the kernel, because this is
+/// just a base-length pair --- taken by itself, it doesn't let you actually
+/// _access_ the memory.
+///
+/// # Invariants
+///
+/// `s.is_empty()` implies `s.base_addr() == s.end_addr()`, and vice versa.
+///
+/// `s.base_addr() <= s.end_addr()` must hold.
+pub trait UserSlice {
+    /// Checks whether the slice spans zero bytes. Empty slices are opted out of
+    /// access checking to allow task code to use literals like `&[]`.
+    ///
+    /// This must be consistent with the base/end addr implementations, such
+    /// that `is_empty <==> base_addr == end_addr`.
+    fn is_empty(&self) -> bool;
```

Comment:
I could be convinced, but, I'm kind of concerned about default implementations here accidentally making their way into the kernel. This is why I haven't provided any.

As for rephrasing it in terms of base+len... that would accurately describe the types in the kernel, though the kernel's `end_addr` routine has been optimized to avoid the checked arithmetic, so we'd still need to be able to provide an `end_addr` override somehow, and I think that winds up meaning a default impl.